### PR TITLE
[jit] fix tensor.data attribute to not requires_grad

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25350 [jit] fix tensor.data attribute to not requires_grad**

Summary:
This fix the .data attribute where it should not require_grad for its returned value and use whatever python use to get the data with version_counter=0 always

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25350

Differential Revision: [D17100829](https://our.internmc.facebook.com/intern/diff/D17100829)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -612,6 +612,22 @@ def forward(self, input):
         m_dropout.eval()
         self.assertEqual(dropout(input) + 1, m_dropout(input))
 
+    def test_script_grad_mode(self):
+        @torch.jit.script
+        def test_set_grad_mode_function(x):
+            # type: (Tensor) -> Tuple[Tensor, Tensor]
+            torch.set_grad_enabled(True)
+            y = x * 2
+            torch.set_grad_enabled(False)
```
Comment (zdevito) [inline test/test_jit.py]: This is going to leave gradients off after this test, right? That seems like it will break other tests.

File: test/test_jit.py [hunk 2]
```diff
@@ -612,6 +612,22 @@ def forward(self, input):
...
+            z = x * 2
+            return y, z
+
+        x = torch.randn(2, 2, requires_grad=True)
+        print(test_set_grad_mode_function.graph)
```
Comment (zdevito) [inline test/test_jit.py]: stray print

File: torch/csrc/jit/register_prim_ops.cpp [hunk 3]
```diff
@@ -650,6 +655,13 @@ RegisterOperators reg(
            return 0;
          },
          aliasAnalysisFromSchema()),
+     Operator(
+         "aten::set_grad_enabled(bool mode) -> ()",
```
Comment (zdevito) [inline torch/csrc/jit/register_prim_ops.cpp]: This needs to be marked as conservative. Otherwise the compiler can move compilation above/below this function during optimizations.

**Review by zdevito — COMMENTED** [review]
A couple comments inline. Also, I would appreciate some context in the PR description. It is currently blank.

File: test/test_jit.py (hunk 1 above)
Comment (wanchaol) [reply test/test_jit.py]: hmm yeah you are right, there's no API to get the current grad_mode so it's hard to restore the environment as before. We should actually enable the context manager version first instead of acting like a function. I will postpone adding this until we have context manager supported.

**Review by zdevito — APPROVED** [review]
